### DISCUSSION TITLE
PageVariable.SetRecord() cannot use a temporary record for the Record parameter.

### DISCUSSION THREAD

**Discussion by christianbraeunlich** [StefanMaron/BusinessCentral.LinterCop#611] [New Rule]
Hello there,

the goal of this rule is to deal with the record parameter of the PageVariable.SetRecord() procedure which cannot use a temporary record for the Record parameter.

Stated in the docs (see: https://learn.microsoft.com/de-de/dynamics365/business-central/dev-itpro/developer/methods-auto/page/page-setrecord-method#parameters): "Record: *The record to set as the current record. **You cannot use a temporary record for the Record parameter**.*"

In addition, I sometimes see that this method is used to set the entire table of the page. However, this is not the case because "*You can use this method to set the record to display when the user opens the page.*".

Any additional thoughts on this? Is there already a rule I haven't found yet?

**Comment by Arthurvdv** [comment]
Valid point you're making. If I understand correctly, it's kinda like the (simplefied) scenario below? This indeed doesn't work as expected and returns another customer records it seems.

It seems there's no rule in one of the Code Analyzers for this. Seems doable to create an rule for this.

```AL
pageextension 50100 "Customer List" extends "Customer List"
{
    actions
    {
        addafter(WordTemplate)
        {
            action(SetRecordWithTempTable)
            {
                ApplicationArea = All;
                Caption = 'Set Record with TempTable';

                trigger OnAction()
                var
                    TempCustomer: Record Customer temporary;
                    CustomerCard: Page "Customer Card";
                begin
                    TempCustomer.Init();
                    TempCustomer."No." := '12345';
                    TempCustomer."Name" := 'Set Record with TempTable';
                    TempCustomer.Insert(false);

                    CustomerCard.SetRecord(TempCustomer);
                    CustomerCard.Run();
                end;

            }
        }
    }
}
```

Regarding your addition, about passing an entire table, I'm assuming the best pratice would be that the table variable contains one (single) record? That's easy to determine in runtime, but difficult to determine in the compiler.

**Reply by Arthurvdv** [reply]
Follow-up question: Does this also apply for other operators, like `GetRecord`, `SetSelectionFilter`, `SetTableView`?

**Reply by MBE-NAV** [reply]
Passing the entire table as TEMP is on one hand very attractive in regards of showing only selected values which were calculated before, bcs you don't have to reload them from the SQL database.

Interesting would be the behaviour of the Pages in newer BC Versions as they do not load all records, but only a few more than the ones which are displayed.
I can imagine this being a challenge when passing the whole table as temporary, as those temporary records then need to stay in RAM and cannot be reloaded on the fly from the SQL db during scrolling.

Might cause very slow system reactions as a result.

**Comment by frottke** [comment]
I would like to add our support to the suggestion already mentioned. A rule that identifies cases where a temporary record is passed to PageVariable.SetRecord() would not only improve code quality but also help prevent unintended data loss, which makes this enhancement particularly valuable for us. We would appreciate it if this could be taken into consideration.

**Comment by DarleenN** [comment]
I support this as well. @frottke nailed it
